donkeycar/parts/oak_d_camera.py

```diff
@@ -0,0 +1,359 @@
+        monoRight.setResolution(dai.MonoCameraProperties.SensorResolution.THE_400_P)
+        monoLeft.setResolution(dai.MonoCameraProperties.SensorResolution.THE_400_P)
+
+        stereo_manip.initialConfig.setCropRect(topLeft.x, topLeft.y, bottomRight.x, bottomRight.y)
+        # manip.setMaxOutputFrameSize(monoRight.getResolutionHeight()*monoRight.getResolutionWidth()*3)
+        stereo.setDefaultProfilePreset(dai.node.StereoDepth.PresetMode.HIGH_DENSITY)
+
+        # Linking
+        # configIn.out.link(manip.inputConfig)
+        monoRight.out.link(stereo.right)
+        monoLeft.out.link(stereo.left)
+        stereo.depth.link(stereo_manip.inputImage)
+        stereo_manip.out.link(xout_depth.input)
+
+    def create_obstacle_dist_pipeline(self):
+
+        # Define sources and outputs
+        monoLeft = self.pipeline.create(dai.node.MonoCamera)
+        monoRight = self.pipeline.create(dai.node.MonoCamera)
+        stereo = self.pipeline.create(dai.node.StereoDepth)
+        spatialLocationCalculator = self.pipeline.create(dai.node.SpatialLocationCalculator)
+
+        # xoutDepth = self.pipeline.create(dai.node.XLinkOut)
+        xoutSpatialData = self.pipeline.create(dai.node.XLinkOut)
+        xinSpatialCalcConfig = self.pipeline.create(dai.node.XLinkIn)
+
+        # xoutDepth.setStreamName("depth")
+        xoutSpatialData.setStreamName("spatialData")
+        xinSpatialCalcConfig.setStreamName("spatialCalcConfig")
+
+        # Properties
+        monoLeft.setResolution(dai.MonoCameraProperties.SensorResolution.THE_400_P)
+        monoLeft.setBoardSocket(dai.CameraBoardSocket.LEFT)
+        monoRight.setResolution(dai.MonoCameraProperties.SensorResolution.THE_400_P)
+        monoRight.setBoardSocket(dai.CameraBoardSocket.RIGHT)
+
+        stereo.setDefaultProfilePreset(dai.node.StereoDepth.PresetMode.HIGH_DENSITY)
+        stereo.setLeftRightCheck(True)
+        stereo.setExtendedDisparity(True)
+        spatialLocationCalculator.inputConfig.setWaitForMessage(False)
+
+        for i in range(4):
+            config = dai.SpatialLocationCalculatorConfigData()
+            config.depthThresholds.lowerThreshold = 200
+            config.depthThresholds.upperThreshold = 10000
+            # 30 - 40 est le mieux
+            config.roi = dai.Rect(dai.Point2f(i*0.1+0.3, 0.35), dai.Point2f((i+1)*0.1+0.3, 0.43))
+            spatialLocationCalculator.initialConfig.addROI(config)
+            # 4 zones
+            # PCLL PCCL PCCR PCRR
+            # -.75 -.75 +.75 +.75
+            
+        # Linking
+        monoLeft.out.link(stereo.left)
+        monoRight.out.link(stereo.right)
+
+        # spatialLocationCalculator.passthroughDepth.link(xoutDepth.input)
+        stereo.depth.link(spatialLocationCalculator.inputDepth)
+
+        spatialLocationCalculator.out.link(xoutSpatialData.input)
+        xinSpatialCalcConfig.out.link(spatialLocationCalculator.inputConfig)
+
+
+    def run(self):
```

Comment:
This is not correct. The `run` method should return the same data as `run_threaded` if the part is called as a non-threaded part. However, if you designed the part to be threaded, then it is fine to not have the `run` method at all. The `update` method runs in a separate thread and in general updates member variables which are then used by the `run_threaded` method. So in case you want to implement the `run` method as well, it would sense to do it by filling these members accordingly and then returning them the same way as `run_threaded`, such that everything happens within the `run` function itself.